tests/app/main/views/test_search.py

```diff
@@ -0,0 +1,67 @@
+import pytest
+from lxml import html
+
+from ...helpers import LoggedInApplicationTest
+
+
+class TestSearchView(LoggedInApplicationTest):
+    @pytest.mark.parametrize("role,expected_code", [
+        ("admin", 200),
+        ("admin-ccs-category", 200),
+        ("admin-ccs-sourcing", 200),
+        ("admin-manager", 403),
+        ("admin-framework-manager", 200),
+    ])
+    def test_find_suppliers_and_services_page_is_only_accessible_to_specific_user_roles(
+        self, role, expected_code
+    ):
+        self.user_role = role
+        response = self.client.get('/admin/find-suppliers-and-services')
+        actual_code = response.status_code
+        assert actual_code == expected_code, "Unexpected response {} for role {}".format(actual_code, role)
+
+    @pytest.mark.parametrize("role, expected_header", [
+        ("admin", "Edit supplier accounts or view services"),
+        ("admin-ccs-category", "Edit suppliers and services"),
+        ("admin-ccs-sourcing", "Edit supplier declarations"),
+        ("admin-framework-manager", "View suppliers and services"),
+    ])
+    def test_should_show_find_suppliers_and_services_page_with_role_appropriate_header(self, role, expected_header):
+        self.user_role = role
+        response = self.client.get('/admin/find-suppliers-and-services')
+        page_html = response.get_data(as_text=True)
+        document = html.fromstring(page_html)
+        header = document.xpath(
+            '//header[@class="page-heading page-heading-without-breadcrumb"]//h1/text()')[0].strip()
+
+        assert response.status_code == 200
+        assert header == expected_header
+
+    @pytest.mark.parametrize(
+        "role, supplier_name_form_should_exist, supplier_duns_form_should_exist, service_id_form_should_exist", [
+            ("admin", True, True, True),
+            ("admin-ccs-category", True, True, True),
+            ("admin-ccs-sourcing", True, True, False),
+            ("admin-framework-manager", True, True, True),
+        ])
+    def test_forms_visible_for_relevant_roles(
+            self, role, supplier_name_form_should_exist, supplier_duns_form_should_exist, service_id_form_should_exist
+    ):
+        self.user_role = role
+        response = self.client.get('/admin/find-suppliers-and-services')
+        document = html.fromstring(response.get_data(as_text=True))
+        question_headings = [heading.strip() for heading in document.xpath('//span[@class="question-heading"]/text()')]
+
+        # We test that we can find the relevant question headings as a proxy for finding the forms themselves
+        assert len(question_headings) == (
+            supplier_name_form_should_exist + supplier_duns_form_should_exist + service_id_form_should_exist
+        )
```

Comment:
👍 👍 👍 Love it

Reply:
Ta - I was quietly pleased with this myself.